### ISSUE TITLE
SymPy translation to JavaScript

### ISSUE THREAD

**Issue by Kreijstal** [open-source-ideas/open-source-ideas#165]
## Project description

As far as I know there are only three JavaScript CASes with various issues, these are
- MathJS
- Algebrite
- Nerdamer

However none of those are at the level of sympy. So I figured out there could be a translation of sympy into JavaScript and there are many paths to realize this.

1.  Translate NumPy into JavaScript by hand
    - I like this one a lot, however it needs a lot of time, and manpower, and you would finally have a native good quality CAS on JavaScript, the cons are, you would have to mantain it, and update it at the same time as sympy.
2.  Use [transcription software](https://github.com/QQuick/Transcrypt) to transcribe python into JavaScript
    - This might be possible but I don't know if it works well for huge projects. 
3.  Just compile Python into JavaScript with emscripten.
    - This would mean that you wouldn't have to spend so much time mantaining your port of Sympy, but depending on how you compiled it, it will take a really long time to load your libraries and might not work on all browsers (This one is what pyodide has done but currently there is no module on npm available)
4.  Use a Python intepreter in JavaScript
    - I think that is what [brython](https://brython.info/) has done

[_Write what technology is relevant. What language, what platform, any particular library/framework/existing project it is based on?_]
[ Convert sympy to javascript](https://groups.google.com/forum/#!topic/sympy/7HXq4YbzvMo)
[Pyodide](https://github.com/iodide-project/pyodide)
## Complexity and required time

[_Please only tick off one box in each category by changing `[ ]` to `[x]`. The labels on the project will then be updated by the maintainers as soon as possible._]

### Complexity

- [x] Beginner - This project requires no or little prior knowledge of the technolog(y|ies) specified to contribute to the project
- [ ] Intermediate - The user should have some prior knowledge of the technolog(y|ies) to the point where they know how to use it, but not necessarily all the nooks and crannies of the technology
- [ ] Advanced - The project requires the user to have a good understanding of all components of the project to contribute

### Required time (ETA)

- [ ] Little work - A couple of days
- [ ] Medium work - A week or two
- [x] Much work - The project will take more than a couple of weeks and serious planning is required

**Comment by jarble** [comment]
Alternatively, it might be possible to automatically translate some Sympy functions to JavaScript using [JS2Py](https://github.com/PiotrDabkowski/Js2Py).

**Comment by stefnotch** [comment]
I would like to point out that this is already sort of a thing using [pyodide](https://github.com/iodide-project/pyodide/), which fully supports sympy! 🎉

**Comment by hodovani** [comment]
@Kreijstal isn't it easier to extend an existed lib than create a new one? It might be faster to add set theory, simplify boolean expressions, solve inequalities to Math.JS. What do you think?

**Comment by Kreijstal** [comment]
> @Kreijstal isn't it easier to extend an existed lib than create a new one? It might be faster to add set theory, simplify boolean expressions, solve inequalities to Math.JS. What do you think?

I think currently, it really makes little difference. Since the amount of effort to make an alternative to sympy is around the same, but of course, if you decide to update existing math libraries, then, by all means. Please do so. Also it doesn't mean we cannot use their functions that they define if it saves time.

**Comment by hodovani** [comment]
@Kreijstal do you know if someone worked on this feature? I would like to learn more about these algorithms by doing tools.

**Comment by Kreijstal** [comment]
In order to do inequalities you need to have a way of using operation on sets of numbers see this for example 
https://github.com/jiggzson/nerdamer/issues/456
https://github.com/jiggzson/nerdamer/issues/350#issuecomment-468919997_
of course these are just implementations of things that have been done on sympy, there are of course many CAS (Computer Algebra systems, see matlab, mathematica) But of course, they're not open source, and not supported on the browser.

**Comment by stefnotch** [comment]
I checked out the different computer algebra systems for [QuantumSheet (WIP - searching for collaborators)](https://github.com/stefnotch/quantum-sheet). Basically, this is what I found

> Mathematics Backends
> - [ ] [SymPy](https://github.com/iodide-project/pyodide/)
> - [ ] [Giac/Xcas](https://github.com/brentan/emgiac)
> 
> Less Important
> - [ ] [Yacas](http://www.yacas.org/yacas_online/yacas_online.html)
> - [ ] [GAP](https://github.com/gap-system/gap/issues/2292#issuecomment-647495335)
> 
> Far future
> - [ ] [Maxima](https://sourceforge.net/p/maxima/feature-requests/)
> - [ ] [Fricas](https://github.com/fricas/fricas/issues/16)

**Comment by Kreijstal** [comment]
> I would like to point out that this is already sort of a thing using [pyodide](https://github.com/iodide-project/pyodide/), which fully supports sympy!

I mean, sure I've done this and it's pretty nice, but of course it's a bit heavy since it has to use nasm, and painfully slow, on phones, for example, would be nice native javascript but of course, its still an option.

**Comment by stefnotch** [comment]
Yes, the size of pyodide is definitely a big issue. It should be possible to slash the size at least in half, as documented here https://github.com/iodide-project/pyodide/issues/646

If you look at the network tab, you can see that it transfers multiple megabytes of data (especially on first load, on subsequent loads most of it should be cached)
![image](https://user-images.githubusercontent.com/10220080/105213058-7eed7500-5b4e-11eb-8e0c-8819b7106aba.png)

Though, for development purposes, this is actually not an issue. If you use a shared worker and make sure to keep one tab open, you can keep the worker alive instead of reloading it every time.

**Comment by hodovani** [comment]
> In order to do inequalities you need to have a way of using operation on sets of numbers see this for example
> [jiggzson/nerdamer#456](https://github.com/jiggzson/nerdamer/issues/456)
> [jiggzson/nerdamer#350 (comment)](https://github.com/jiggzson/nerdamer/issues/350#issuecomment-468919997)_
> of course these are just implementations of things that have been done on sympy, there are of course many CAS (Computer Algebra systems, see matlab, mathematica) But of course, they're not open source, and not supported on the browser.

sorry, I am not sure that I understand your message. Do you know if we have something that is similar to SymPy? If not. Should we start creating this?

**Comment by liudonghua123** [comment]
If we want to implement some features like python's `sympy`, the operator overload should be necessary.

**Comment by stefnotch** [comment]
> If we want to implement some features like python's `sympy`, the operator overload should be necessary.

What do you mean? If you're referring to Javascript operator overloading, then no, that's not necessary at all.

**Comment by stefnotch** [comment]
> 4. Use a Python intepreter in JavaScript
>    I think that is what brython has done

Regarding point 4, the Brython developer has attempted that and ran into a few blocking issues that would have to be sorted out on the Sympy side
https://github.com/sympy/sympy/issues/18889

**Comment by liudonghua123** [comment]
> > If we want to implement some features like python's `sympy`, the operator overload should be necessary.
> 
> What do you mean? If you're referring to Javascript operator overloading, then no, that's not necessary at all.

If there isn't operator overloading, some arithmetic operation like the following code would implement and write using a complicate method.

```python
# this is a sample python sympy code
x = symbols('x')
expr = x**2 + 2*x + 1
```

The correspond js code maybe like this

```js
import symbols, {add, multiply, exp} from 'symbols';
const x = symbols('x')
const expr = add(add(exp(x, 2), multiply(x, 2), 1)
```

The form of `const expr = add(add(exp(x, 2), multiply(x, 2), 1)` is too bad.

**Comment by stefnotch** [comment]
@liudonghua123 As far as I know, you would need more than just [operator overloading](https://github.com/tc39/proposal-operator-overloading). You would need something closer to expression trees.

For example, take the following expression (Javascript)
```js
const expr = 3/7;
```
Since it's a fraction that can't be represented in base 2 nor in base 10, you would want to store the actual fraction `3 / 7` and not the result (`0.4285...`).

To do that with operator overloading, you have to overload the division operator *for the built-in number type*. Javascript might never get the ability to do that.

A more doable approach right now would be parsing string-expressions and turning them into sympy code. Something like
```js
const expr = sympy("3/7"); // Generates an expression tree
```

**Comment by KOLANICH** [comment]
I guess another way to achieve it is to compile https://github.com/symengine/symengine into WASM.
